src/middleware/persist.ts

```diff
@@ -0,0 +1,353 @@
+// Types
+
+type Persist =
+  <T extends UnknownState, U>
+    ( stateInitializer: StateInitializer<T, Store<T>>
+    , options: PersistOptions<T, U>
+    ) =>
+      StateInitializer<T, Store<T> & PersistStore<T, U>>
+
+interface PersistOptions<T extends UnknownState, U>
+  { /** Name of the storage (must be unique) */
+    name: string
+    
+    /**
+     * A function returning a storage.
+     * The storage must fit `window.localStorage`'s api (or an async version of it).
+     * For example the storage could be `AsyncStorage` from React Native.
+     *
+     * @default () => localStorage
+     */
+  , getStorage?: () => PersistentStorage
+
+    /**
+     * Use a custom serializer.
+     * The returned string will be stored in the storage.
+     *
+     * @default JSON.stringify
+     */
+    serialize?:
+      (storageValue: PersistentStorageValue<U>) => MaybePromise<string>
+
+    /**
+     * Use a custom deserializer.
+     * Must return an object matching PersistentStorageValue<State>
+     *
+     * @param serializedString The storage's current value.
+     * @default JSON.parse
+     */
+    deserialize?:
+      (serializedString: string) => MaybePromise<PersistentStorageValue<T>>
+
+    /**
+     * Filter the persisted value.
+     *
+     * @param state The state's value
+     */
+    partialize?: (state: T) => U
+    
+    /**
+     * A function returning another (optional) function.
+     * The main function will be called before the state rehydration.
+     * The returned function will be called after the state rehydration or when an error occurred.
+     */
+    onRehydrateStorage?:
+      (state: T) =>
+        void | ((state?: T, error?: unknown) => void)
+
+    /**
+     * If the stored state's version mismatch the one specified here, the storage will not be used.
+     * This is useful when adding a breaking change to your store.
+     */
+    version?: number
+
+    /**
+     * A function to perform persisted state migration.
+     * This function will be called when persisted state versions mismatch with the one specified here.
+     */
+    migrate?:
+      (persistedState: unknown, version: number) => MaybePromise<T>
+
+    /**
+     * A function to perform custom hydration merges when combining the stored state with the current one.
+     * By default, this function does a shallow merge.
+     */
+    merge?:
+      (persistedState: unknown, currentState: T) => T
+  }
+
+interface PersistentStorage
+  { getItem: (name: string) => string | null | Promise<string | null>
+  , setItem: (name: string, value: string) => void | Promise<void>
+  , removeItem: (name: string) => void | Promise<void>
+  }
+
+interface PersistentStorageValue<T>
+  { state: T
+  , version?: number
+  }
+
+interface PersistStore<T extends UnknownState, U>
+  { persist:
+      { setOptions: (options: O.Partial<PersistOptions<T, U>>) => void
+      , clearStorage: () => void
+      , rehydrate: () => Thenable<void>
```

Comment:
This previously was `rehydrate: () => Promise<void>` but we were lying in the types and it's actually is `rehydrate: () => Thenable<void>`.
So previously someone writing `rehydrate().finally()` would have compiled and it would have given a runtime error, but now it won't.